### PR TITLE
Document how to run code coverage with phpunit

### PR DESCRIPTION
**Pull request by tswfi**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 8.x
| Description?  | Document how to run coverage tests with phpunit
| Fixed ticket? |

### CODE REVIEW

File: testing/how-to-launch-tests.md [hunk 1]
```diff
@@ -47,3 +47,25 @@ vendor/bin/phpunit --debug -c tests/Unit/phpunit.xml tests/Unit/PrestaShopBundle
...
 {{% /notice %}}
+
+# Checking for code coverage in phpunit tests
+
+To check if you covered everything in your tests cases its best to run the tests with coverage report.
+
+First you have to enable xdebug. For the docker dev environment run
+
+```
+docker-php-ext-enable xdebug
```
Comment (PierreRambaud) [inline testing/how-to-launch-tests.md]: Why talk about docker? :thinking:

Comment (tswfi) [reply testing/how-to-launch-tests.md]: yeah, maybe this should have a link to this page: https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/contribute_using_docker/#how-to-become-a-core-contributor-using-docker (and that page to have a link back to here)

Comment (PierreRambaud) [reply testing/how-to-launch-tests.md]: and also, it's when you're inside the container, for someone who don't know docker, docker-compose, he will not understand anything about this sentence

**Review by matks — CHANGES_REQUESTED** [review]
Agreed with @PierreRambaud : the devdocs must be as accessible as possible 😉

**Comment by tswfi** [comment]
yup, added few more steps and links to more information

File: testing/how-to-launch-tests.md [hunk 2]
```diff
@@ -48,19 +48,27 @@ vendor/bin/phpunit --debug -c tests/Unit/phpunit.xml tests/Unit/PrestaShopBundle
...
-To check if you covered everything in your tests cases its best to run the tests with coverage report.
+To check if you covered everything in your tests cases its best to run the tests with [phpunit coverage report](https://phpunit.readthedocs.io/en/9.5/code-coverage-analysis.html).
 
-First you have to enable xdebug. For the docker dev environment run
+First get your [docker dev environment up and running]({{< relref "/8/contribute/contribute-pull-requests/contribute_using_docker#install-prestashop-core" >}})
 
+After that is up and running you need to enable xdebug inside the container and run your tests with coverage
+
+```bash
+docker exec -it `docker-compose ps -q prestashop-git` bash
```
Comment (PierreRambaud) [inline testing/how-to-launch-tests.md]: Docker exec to execute docker-compose command? oO

Comment (tswfi) [reply testing/how-to-launch-tests.md]: heh, I was first trying to make that oneliner and didn't succeed and forgot to modify all of it

any idea how to make that oneliner?

File: testing/how-to-launch-tests.md (hunk 1 above)
Comment (tswfi) [reply testing/how-to-launch-tests.md]:
```
docker-compose exec prestashop-git XDEBUG_MODE=coverage vendor/bin/phpunit --coverage-text -c tests/Unit/phpunit.xml tests/Unit/PrestaShopBundle/Command/ConfigCommandTest.php 
```

would be nice but I cant get it working (probably just something stupid but I cannot see it)

```
OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "XDEBUG_MODE=coverage": executable file not found in $PATH: unknown
```

Comment (tswfi) [reply testing/how-to-launch-tests.md]: aah, of course .. the env variable needs to be pushed with -e

```
docker-compose exec -e XDEBUG_MODE=coverage prestashop-git vendor/bin/phpunit --coverage-text -c tests/Unit/phpunit.xml tests/Unit/PrestaShopBundle/Command/ConfigCommandTest.php 
```

seems to work correctly

I'll update in a second

Comment (tswfi) [reply testing/how-to-launch-tests.md]: @PierreRambaud see, I'm learning :)

File: testing/how-to-launch-tests.md (hunk 2 above)
Comment (tswfi) [reply testing/how-to-launch-tests.md]: doh, my other comments went to wrong "thread" f054748

**Review by matks — COMMENTED** [review]
I'm afraid I'm not a docker expert so I need @PierreRambaud to review this contribution

**Comment by PierreRambaud** [comment]
> I'm afraid I'm not a docker expert so I need @PierreRambaud to review this contribution

I don't get the relation between docker and code coverage.
Above there is no mention of docker.
Here's is how you manage to make it work for your environment. If you want to explain something it should be easier and better to only talk about `./vendor/bin/phpunit` not the environment you're  working on

**Comment by tswfi** [comment]
docker environment is a "sure" way to get a working environment and running the tests inside the container makes everybody run the tests the same way.

The devdocs has information on how to get the docker environment up: https://devdocs.prestashop.com/8/contribute/contribute-pull-requests/contribute_using_docker/#install-prestashop-core and using the same environment to run the tests is simple for everybody.

One could argue that all the documentation should be done on top of the docker environment to have "one way of doing things" and if someone wants to do something else then they will probably know enough to do it.

**Comment by kpodemski** [comment]
The part I'm missing is:

> If you use Docker env [...]

Not everyone runs docker env, and you can simply have native env prepared. Maybe we should make sure to mention it?

**Review by MeKeyCool — APPROVED** [review]
> The part I'm missing is:
> 
> > If you use Docker env [...]
> 
> Not everyone runs docker env, and you can simply have native env prepared. Maybe we should make sure to mention it?

Sure but documentation is not the "only" way to use the project.
I think it is interesting to add some documentation about test examples.

It would be great to have an example of each test type we can find in github actions (unit, phpstan, ...; https://github.com/PrestaShop/PrestaShop/tree/develop/.github/workflows) using different setup (docker, local, ...)

@kpodemski , would you prefer to have two separated sections one for different install and setup examples, the other one for test command list ? :thinking:

**Comment by matks** [comment]
Hello @thomasnares here is an interesting question: should this content be added to doc or not ? 😅 we could not find the answer in 2021

**Comment by thomasnares** [comment]
I would totally add content about testing code coverage, but not on how to run it with Docker. 
Docker or native doesn't really differ:  I would mention that you need xdebug, and enable `XDEBUG_MODE=coverage`.

**Comment by thomasnares** [comment]
i proposed this update : https://github.com/tswfi/docs/pull/31

**Comment by thomasnares** [comment]
This one can be closed

**Pull request closed without merging**